torch/csrc/LazyTensor.cpp

```diff
@@ -0,0 +1,385 @@
+ * any shape dependent code is run.  The implication is that some core modules,
+ * such as nn.BatchNorm, cannot be executed lazily because their implementations
+ * check shapes in Python.
+ *
+ * Note: LazyTensor also accumulates a graph during calls to `backward`.  Shape
+ * information is required for error checking during backward propagation,
+ * but with LazyTensors we skip these checks.
+ *
+ * Note: LazyTensor assumes all operations are re-computable!  This assumption
+ * saves us from using too much memory without explicit need.  However, it also
+ * induces potentially unwanted re-compute and BREAKS semantics for operations
+ * that require state.  TODO: blacklist broken ops
+ *
+ * Note: LazyTensors cannot be used in multi-threaded environment.  This paradigm
+ * should subvert the need for parallelism at this level anyway.
+ *
+ */
+
+namespace at {
+
+// @returns pair of output and required stack inputs
+std::pair<torch::jit::Node*, std::vector<const LazyTensorImpl*>> getRequiredGraphValues(
+    const LazyTensorImpl* lt_,
+    std::shared_ptr<torch::jit::Graph> g) {
+  // Convert LazyTensorImpl to node in the graph, inserting before
+  // the "dep" node.
+  auto createNode = [&g](
+                     const LazyTensorImpl* lt,
+                     torch::jit::Node* dep) -> torch::jit::Node* {
+    auto fs = lt->schema();
+    auto s = Symbol::fromQualString(fs.name());
+    auto* node = g->create(s, fs.returns().size());
+    if (dep) {
+      g->setInsertPoint(dep);
+    }
+    node = g->insertNode(node);
+    return node;
+  };
+
+  // Setup output of the whole graph
+  auto* root_node = createNode(lt_, nullptr);
+  torch::jit::Value* output = root_node->outputs()[lt_->outputIndex()];
+  std::vector<const LazyTensorImpl*> inputs;
+
+  // Create frontier for resolving deps
+  // The form is { IValue, Node (that depends on IValue) }
+  std::vector<std::pair<c10::IValue, torch::jit::Node*>> frontier;
+  for (const auto& input : lt_->inputs()) {
+    frontier.emplace_back(std::make_pair(input, root_node));
+  }
+
+  while (frontier.size()) {
+    std::vector<std::pair<c10::IValue, torch::jit::Node*>> next_frontier;
+    for (auto dep_pair : frontier) {
+      auto ival = dep_pair.first;
+      auto dep_node = dep_pair.second;
+
+      // Is the IValue a LazyTensor?
+      if (ival.isTensor() && ival.toTensor().is_lazy()) {
+        auto t = ival.toTensor();
+        auto* lt = static_cast<LazyTensorImpl*>(t.unsafeGetTensorImpl());
+        // Is lt just storing a tensor? This is an input
+        if (lt->isResolved()) {
+          auto* v = g->addInput();
+          inputs.emplace_back(lt);
+          dep_node->addInput(v);
+        } else {
+          auto* node = createNode(lt, dep_node);
+          dep_node->addInput(node->outputs()[lt->outputIndex()]);
+          for (const auto& input : lt->inputs()) {
+            next_frontier.emplace_back(std::make_pair(input, node));
```

Comment:
You are doing the book-keeping externally without a hash map, which means you're doing a tree traversal rather than a graph traversal. This looks dangerous. What if I have:

```
x1 = x0.to_lazy()
x2 = x1 + x1
x3 = x2 + x2
# ...
```

It seems to me that you would end up with an exponentially sized graph in this case.